sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs

```diff
@@ -165,17 +171,45 @@ private ProgramState VisitAssignmentExpression(AssignmentExpressionSyntax assign
                     programState.HasConstraint(binderSymbolValue, ObjectConstraint.Null))
                 {
                     // The formatter is considered unsafe if the binder is null.
-                    return programState.SetConstraint(formatterSymbolValue, SerializationBinderConstraint.Unsafe);
+                    return programState.SetConstraint(formatterSymbolValue, SerializationConstraint.Unsafe);
                 }
 
                 var binderType = semanticModel.GetTypeInfo(assignmentExpression.Right).Type;
-                var constraint = IsBinderSafe(binderType)
-                    ? SerializationBinderConstraint.Safe
-                    : SerializationBinderConstraint.Unsafe;
+                var constraint = IsTypeSafe(binderType)
+                    ? SerializationConstraint.Safe
+                    : SerializationConstraint.Unsafe;
 
                 return programState.SetConstraint(formatterSymbolValue, constraint);
             }
 
+            private bool HasSafeResolverInitialization(ObjectCreationExpressionSyntax objectCreation)
+            {
+                // JavaScriptSerializer has 2 constructors:
+                // - JavaScriptSerializer(): unsafe since it doesn't give the option to set a provider
+                // - JavaScriptSerializer(JavaScriptTypeResolver): this one is safe only if the given resolver is safe
+                // See: https://docs.microsoft.com/en-us/dotnet/api/system.web.script.serialization.javascriptserializer.-ctor?view=netframework-4.8
+
+                if (objectCreation.ArgumentList.Arguments.Count == 0)
+                {
+                    return false;
+                }
+
+                var resolverType = objectCreation.ArgumentList.Arguments[0];
+                if (resolverType.Expression.IsNullLiteral())
+                {
+                    return false;
+                }
+
+                var typeSymbol = semanticModel.GetTypeInfo(resolverType.Expression).Type;
+                // If the symbol cannot be determined we consider the resolver safe to avoid FPs.
+                if (typeSymbol == null)
+                {
+                    return true;
+                }
+
+                return !typeSymbol.Is(KnownType.System_Web_Script_Serialization_SimpleTypeResolver) && IsTypeSafe(typeSymbol);
```

Comment:
```suggestion
                return typeSymbol == null || (!typeSymbol.Is(KnownType.System_Web_Script_Serialization_SimpleTypeResolver) && IsTypeSafe(typeSymbol));
```
Why is this logic different than for `HasSafeBinderInitialization` ? For `HasSafeBinderInitialization` if `typeSymbol` is null, method will return false.